src/core/processing/processingalgorithm.cpp

```diff
@@ -0,0 +1,271 @@
+  }
+}
+
+void ProcessingAlgorithm::setInPlaceFeatures( const QList<QgsFeature> &features )
+{
+  if ( mInPlaceFeatures == features )
+  {
+    return;
+  }
+
+  mInPlaceFeatures = features;
+
+  emit inPlaceFeaturesChanged();
+
+  if ( mPreview && mAlgorithm )
+  {
+    run( true );
+  }
+}
+
+void ProcessingAlgorithm::setParameters( const QVariantMap &parameters )
+{
+  if ( mAlgorithmParameters == parameters )
+  {
+    return;
+  }
+
+  mAlgorithmParameters = parameters;
+
+  emit parametersChanged();
+
+  if ( mPreview && mAlgorithm )
+  {
+    run( true );
+  }
+}
+
+void ProcessingAlgorithm::setPreview( bool preview )
+{
+  if ( mPreview == preview )
+  {
+    return;
+  }
+
+  mPreview = preview;
+
+  emit previewChanged();
+
+  if ( mPreview && mAlgorithm )
+  {
+    run( true );
+  }
+  else
+  {
+    if ( !mPreviewGeometries.isEmpty() )
+    {
+      mPreviewGeometries.clear();
+      emit previewGeometriesChanged();
+    }
+  }
+}
+
+bool ProcessingAlgorithm::run( bool previewMode )
+{
+  if ( !mAlgorithm )
+  {
+    return false;
+  }
+
+  if ( previewMode )
+  {
+    mPreviewGeometries.clear();
+    emit previewGeometriesChanged();
+  }
+
+  QgsProcessingContext context;
+  context.setFlags( QgsProcessingContext::Flags() );
+  context.setProject( QgsProject::instance() );
+
+  QgsProcessingFeedback feedback;
+  context.setFeedback( &feedback );
+
+  QVariantMap parameters = mAlgorithmParameters;
+
+  if ( mInPlaceLayer )
+  {
+    if ( mInPlaceFeatures.isEmpty() || !mAlgorithm->supportInPlaceEdit( mInPlaceLayer.data() ) )
+    {
+      return false;
+    }
+
+    context.expressionContext().appendScope( mInPlaceLayer->createExpressionContextScope() );
+
+    QStringList featureIds;
+    for ( const QgsFeature &feature : mInPlaceFeatures )
+    {
+      featureIds << QString::number( feature.id() );
+    }
+
+    const QgsProcessingFeatureBasedAlgorithm *featureBasedAlgorithm = dynamic_cast<const QgsProcessingFeatureBasedAlgorithm *>( mAlgorithm );
+    if ( featureBasedAlgorithm )
+    {
+      parameters[featureBasedAlgorithm->inputParameterName()] = QgsProcessingFeatureSourceDefinition( mInPlaceLayer->id(),
+                                                                                                      false,
+                                                                                                      -1,
+                                                                                                      Qgis::ProcessingFeatureSourceDefinitionFlags(),
+                                                                                                      Qgis::InvalidGeometryCheck(),
+                                                                                                      QStringLiteral( "$id IN (%1)" ).arg( featureIds.join( ',' ) ) );
+      parameters[QStringLiteral( "OUTPUT" )] = QStringLiteral( "memory:" );
+
+      if ( !previewMode )
+      {
+        mInPlaceLayer->startEditing();
```

Comment:
You might consider mInPlaceLayer->beginEditCommand() here, if you want all the operations to undo-able in a single step

Reply:
@nyalldawson , we don't use QGIS' internal buffer, so that's not needed ATM. But, that makes me think our own history handling should have a command-like concept to store multiple changes.